File: CallX2/app/src/main/java/com/callx/app/adapters/MessagePagingAdapter.java
package com.callx.app.adapters;

import android.content.Context;
import android.content.Intent;
import android.media.MediaPlayer;
import android.view.*;
import android.widget.*;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.paging.PagingDataAdapter;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;
import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.callx.app.R;
import com.callx.app.activities.MediaViewerActivity;
import com.callx.app.models.Message;
import com.callx.app.utils.FileUtils;
import com.callx.app.utils.MediaCache;

import java.text.SimpleDateFormat;
import java.util.Locale;

/**
 * MessagePagingAdapter — Paging 3 PagingDataAdapter for chat messages.
 *
 * Drop-in replacement for MessageAdapter when loading messages from Room DB
 * via Pager3 + PagingSource. Supports sent/received layout types, text,
 * image, audio, file, and video message rendering identical to MessageAdapter.
 *
 * Usage in ChatActivity:
 *   MessagePagingAdapter pagingAdapter = new MessagePagingAdapter(uid, false);
 *   binding.rvMessages.setAdapter(pagingAdapter);
 *   viewModel.getPagedMessages(chatId).observe(this, pagingAdapter::submitData);
 */
public class MessagePagingAdapter
        extends PagingDataAdapter<Message, MessagePagingAdapter.VH> {

    // ── DiffUtil — required by PagingDataAdapter ──────────────────
    private static final DiffUtil.ItemCallback<Message> DIFF =
        new DiffUtil.ItemCallback<Message>() {
            @Override
            public boolean areItemsTheSame(@NonNull Message a, @NonNull Message b) {
                return a.messageId != null && a.messageId.equals(b.messageId);
            }

            @Override
            public boolean areContentsTheSame(@NonNull Message a, @NonNull Message b) {
                return a.messageId.equals(b.messageId)
                    && safeEquals(a.text, b.text)
                    && safeEquals(a.type, b.type)
                    && safeEquals(a.status, b.status)
                    && a.timestamp == b.timestamp
                    && a.edited == b.edited
                    && a.deleted == b.deleted;
            }

            private boolean safeEquals(String x, String y) {
                if (x == null && y == null) return true;
                if (x == null || y == null) return false;
                return x.equals(y);
            }
        };

    // ── View types ────────────────────────────────────────────────
    private static final int TYPE_SENT     = 1;
    private static final int TYPE_RECEIVED = 2;

    // ── Fields ────────────────────────────────────────────────────
    private final String currentUid;
    private final boolean isGroup;
    private final SimpleDateFormat timeFmt =
            new SimpleDateFormat("hh:mm a", Locale.getDefault());

    private ActionListener actionListener;
    private MediaPlayer player;
    private int playingPos = -1;

    // ── Interface for long-press actions ─────────────────────────
    public interface ActionListener {
        void onReply(Message m);
        void onDelete(Message m);
        void onReact(Message m, String emoji);
        void onStar(Message m);
        void onCopy(Message m);
        void onForward(Message m);
    }

    public MessagePagingAdapter(String currentUid, boolean isGroup) {
        super(DIFF);
        this.currentUid = currentUid;
        this.isGroup    = isGroup;
    }

    public void setActionListener(ActionListener l) {
        this.actionListener = l;
    }

    // ──────────────────────────────────────────────────────────────
    @Override
    public int getItemViewType(int position) {
        Message m = getItem(position);
        if (m == null) return TYPE_RECEIVED;
        return currentUid.equals(m.senderId) ? TYPE_SENT : TYPE_RECEIVED;
    }

    @NonNull
    @Override
    public VH onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        int layout = viewType == TYPE_SENT
                ? R.layout.item_message_sent
                : R.layout.item_message_received;
        View v = LayoutInflater.from(parent.getContext()).inflate(layout, parent, false);
        return new VH(v);
    }

    @Override
    public void onBindViewHolder(@NonNull VH h, int position) {
        Message m = getItem(position);
        if (m == null) {
            // Placeholder — show shimmer or empty
            h.tvMessage.setVisibility(View.GONE);
            return;
        }
        bindMessage(h, m, position);
    }

    // ──────────────────────────────────────────────────────────────
    // Core bind logic (mirrors MessageAdapter)
    // ──────────────────────────────────────────────────────────────
    private void bindMessage(@NonNull VH h, @NonNull Message m, int position) {
        Context ctx = h.itemView.getContext();
        boolean sent = currentUid.equals(m.senderId);

        // Reset visibility
        h.tvMessage.setVisibility(View.GONE);
        if (h.ivImage    != null) h.ivImage.setVisibility(View.GONE);
        if (h.llAudio    != null) h.llAudio.setVisibility(View.GONE);
        if (h.llFile     != null) h.llFile.setVisibility(View.GONE);
        if (h.tvTime     != null) h.tvTime.setVisibility(View.VISIBLE);

        // Timestamp
        if (h.tvTime != null && m.timestamp > 0) {
            h.tvTime.setText(timeFmt.format(new java.util.Date(m.timestamp)));
        }

        // Sender name (group chats)
        if (isGroup && !sent && h.tvSenderName != null) {
            h.tvSenderName.setVisibility(View.VISIBLE);
            String sn = m.senderName != null ? m.senderName : "Member";
            h.tvSenderName.setText(sn);
        } else if (h.tvSenderName != null) {
            h.tvSenderName.setVisibility(View.GONE);
        }

        // Deleted message
        if (Boolean.TRUE.equals(m.deleted)) {
            h.tvMessage.setVisibility(View.VISIBLE);
            h.tvMessage.setText(sent ? "You deleted this message" : "This message was deleted");
            h.tvMessage.setAlpha(0.6f);
            return;
        }

        // ── Render by type ───────────────────────────────────────
        String type = m.type != null ? m.type : "text";
        switch (type) {
            case "image":
            case "gif":
                if (h.ivImage != null) {
                    h.ivImage.setVisibility(View.VISIBLE);
                    String url = m.mediaUrl != null ? m.mediaUrl : m.text;
                    // FIX v14: Check MediaCache first — agar locally stored hai to seedha load,
                    // zero network. Otherwise Glide se load karo aur background mein cache karo.
                    java.io.File cachedImg = MediaCache.getCached(ctx, url);
                    if (cachedImg != null) {
                        android.util.Log.d("PagingAdapter", "Image cache HIT: " + cachedImg.getName());
                        Glide.with(ctx).load(cachedImg)
                            .diskCacheStrategy(DiskCacheStrategy.ALL)
                            .placeholder(R.drawable.ic_file)
                            .error(R.drawable.ic_file)
                            .into(h.ivImage);
                    } else {
                        android.util.Log.d("PagingAdapter", "Image cache MISS, downloading: " + url);
                        Glide.with(ctx).load(url)
                            .diskCacheStrategy(DiskCacheStrategy.ALL)
                            .placeholder(R.drawable.ic_file)
                            .error(R.drawable.ic_file)
                            .into(h.ivImage);
                        // Background mein cache karo taaki agli baar zero data use ho
                        MediaCache.get(ctx, url, new MediaCache.Callback() {
                            @Override public void onReady(java.io.File file) {
                                android.util.Log.d("PagingAdapter", "Image cached: " + file.getName());
                            }
                            @Override public void onError(String reason) {
                                android.util.Log.w("PagingAdapter", "Image cache failed: " + reason);
                            }
                        });
                    }
                    h.ivImage.setOnClickListener(v -> {
                        Intent i = new Intent(ctx, MediaViewerActivity.class);
                        i.putExtra("url", url);
                        i.putExtra("type", "image");
                        ctx.startActivity(i);
                    });
                }
                break;
            case "video":
                if (h.ivImage != null) {
                    h.ivImage.setVisibility(View.VISIBLE);
                    String vUrl = m.mediaUrl != null ? m.mediaUrl : m.text;
                    // FIX v14: Video thumbnail — MediaCache se check karo
                    java.io.File cachedVid = MediaCache.getCached(ctx, vUrl);
                    if (cachedVid != null) {
                        android.util.Log.d("PagingAdapter", "Video cache HIT: " + cachedVid.getName());
                        Glide.with(ctx).load(cachedVid)
                            .diskCacheStrategy(DiskCacheStrategy.ALL)
                            .placeholder(R.drawable.ic_file)
                            .into(h.ivImage);
                    } else {
                        android.util.Log.d("PagingAdapter", "Video cache MISS, downloading: " + vUrl);
                        Glide.with(ctx).load(vUrl)
                            .diskCacheStrategy(DiskCacheStrategy.ALL)
                            .placeholder(R.drawable.ic_file)
                            .into(h.ivImage);
                        // FIX v14: Video background mein pre-cache karo (MediaStreamCache se partial)
                        com.callx.app.cache.MediaStreamCache.getInstance(ctx)
                            .preloadPartial(vUrl, new com.callx.app.cache.MediaStreamCache.DownloadCallback() {
                                @Override public void onComplete(java.io.File file) {
                                    android.util.Log.d("PagingAdapter", "Video partial cached: " + file.getName());
                                }
                                @Override public void onError(String error) {
                                    android.util.Log.w("PagingAdapter", "Video preload failed: " + error);
                                }
                                @Override public void onProgress(int percent) {}
                            });
                    }
                    h.ivImage.setOnClickListener(v -> {
                        Intent i = new Intent(ctx, MediaViewerActivity.class);
                        i.putExtra("url", vUrl);
                        i.putExtra("type", "video");
                        ctx.startActivity(i);
                    });
                }
                break;
            case "audio":
                if (h.llAudio != null && h.btnPlayPause != null) {
                    h.llAudio.setVisibility(View.VISIBLE);
                    String aUrl = m.mediaUrl != null ? m.mediaUrl : m.text;
                    final int pos = position;
                    h.btnPlayPause.setOnClickListener(v -> toggleAudio(h, aUrl, pos));
                    // FIX v14: Audio preload — MediaStreamCache se pehle 512KB cache karo
                    // Taaki play button press karne par turant start ho, buffer nahi kare
                    java.io.File cachedAudio = MediaCache.getCached(ctx, aUrl);
                    if (cachedAudio == null && aUrl != null && !aUrl.isEmpty()) {
                        com.callx.app.cache.MediaStreamCache.getInstance(ctx)
                            .preloadPartial(aUrl, new com.callx.app.cache.MediaStreamCache.DownloadCallback() {
                                @Override public void onComplete(java.io.File file) {
                                    android.util.Log.d("PagingAdapter", "Audio preloaded: " + file.getName());
                                }
                                @Override public void onError(String error) {}
                                @Override public void onProgress(int percent) {}
                            });
                    }
                } else {
                    // Fallback if no audio layout
                    h.tvMessage.setVisibility(View.VISIBLE);
                    h.tvMessage.setText("Audio message");
                }
                break;
            case "file":
            case "document":
                if (h.llFile != null && h.tvFileName != null) {
                    h.llFile.setVisibility(View.VISIBLE);
                    String fName = m.fileName != null ? m.fileName : "File";
                    h.tvFileName.setText(fName);
                    if (h.btnDownload != null) {
                        String fUrl = m.mediaUrl != null ? m.mediaUrl : m.text;
                        h.btnDownload.setOnClickListener(v -> {
                            // Pehle local cache check karo
                            java.io.File cached = MediaCache.getCached(ctx, fUrl);
                            if (cached != null) {
                                FileUtils.openOrDownload(ctx, cached.toURI().toString(), fName);
                                return;
                            }
                            android.widget.Toast.makeText(ctx, "Downloading…", android.widget.Toast.LENGTH_SHORT).show();
                            MediaCache.get(ctx, fUrl, new MediaCache.Callback() {
                                @Override public void onReady(java.io.File file) {
                                    FileUtils.openOrDownload(ctx, file.toURI().toString(), fName);
                                }
                                @Override public void onError(String reason) {
                                    FileUtils.openOrDownload(ctx, fUrl, fName);
                                }
                            });
                        });
                    }
                } else {
                    h.tvMessage.setVisibility(View.VISIBLE);
                    h.tvMessage.setText(m.fileName != null ? m.fileName : "File");
                }
                break;
            default: // "text", "emoji", etc.
                h.tvMessage.setVisibility(View.VISIBLE);
                String txt = m.text != null ? m.text : "";
                if (Boolean.TRUE.equals(m.edited)) txt += " (edited)";
                h.tvMessage.setText(txt);
                h.tvMessage.setAlpha(1f);
                break;
        }

        // ── Delivery status (sent messages only) ─────────────────
        if (sent && h.tvStatus != null) {
            h.tvStatus.setVisibility(View.VISIBLE);
            String status = m.status != null ? m.status : "sent";
            switch (status) {
                case "seen":
                    h.tvStatus.setText("✓✓");
                    h.tvStatus.setTextColor(0xFF4FC3F7); // blue
                    break;
                case "delivered":
                    h.tvStatus.setText("✓✓");
                    h.tvStatus.setTextColor(0xAAFFFFFF);
                    break;
                default:
                    h.tvStatus.setText("✓");
                    h.tvStatus.setTextColor(0xAAFFFFFF);
                    break;
            }
        } else if (h.tvStatus != null) {
            h.tvStatus.setVisibility(View.GONE);
        }

        // ── Long press — action listener ─────────────────────────
        h.itemView.setOnLongClickListener(v -> {
            if (actionListener != null) showActionBottomSheet(ctx, m);
            return true;
        });
    }

    // ──────────────────────────────────────────────────────────────
    // Audio playback toggle
    // ──────────────────────────────────────────────────────────────
    private void toggleAudio(@NonNull VH h, String url, int position) {
        if (playingPos == position && player != null && player.isPlaying()) {
            player.pause();
            if (h.btnPlayPause != null) h.btnPlayPause.setImageResource(R.drawable.ic_play);
            return;
        }
        if (player != null) {
            try { player.stop(); player.release(); } catch (Exception ignored) {}
            player = null;
        }
        playingPos = position;
        if (h.btnPlayPause != null) h.btnPlayPause.setImageResource(R.drawable.ic_pause);

        // Pehle local cache check — cached hai to seedha play (zero data use)
        java.io.File cached = MediaCache.getCached(h.itemView.getContext(), url);
        if (cached != null) {
            playAudioFromPath(h, cached.getAbsolutePath(), position);
            return;
        }

        // FIX v14: MediaStreamCache use karo — pehle 512KB stream karo (fast start),
        // baaki background mein download hota rahe. User ko buffer nahi karega.
        com.callx.app.cache.MediaStreamCache.getInstance(h.itemView.getContext())
            .preloadPartial(url, new com.callx.app.cache.MediaStreamCache.DownloadCallback() {
                @Override public void onComplete(java.io.File file) {
                    // Partial/full file ready — play from local file (zero buffering)
                    android.util.Log.d("AudioPlay", "MediaStreamCache ready, playing: " + file.getName());
                    playAudioFromPath(h, file.getAbsolutePath(), position);
                }
                @Override public void onError(String error) {
                    // Fallback: stream directly from URL
                    android.util.Log.w("AudioPlay", "MediaStreamCache failed, streaming URL: " + error);
                    playAudioFromPath(h, url, position);
                }
                @Override public void onProgress(int percent) {
                    android.util.Log.v("AudioPlay", "Audio preload: " + percent + "%");
                }
            });
    }

    private void playAudioFromPath(@NonNull VH h, String path, int position) {
        try {
            if (player != null) { try { player.release(); } catch (Exception ignored) {} }
            player = new MediaPlayer();
            
            // Agar local file hai to FileDescriptor se set karo (cache files ke liye)
            // Agar URL hai to directly
            if (path.startsWith("http")) {
                player.setDataSource(path);
            } else {
                java.io.File f = new java.io.File(path);
                if (f.exists()) {
                    try (java.io.FileInputStream fis = new java.io.FileInputStream(f)) {
                        player.setDataSource(fis.getFD());
                    }
                } else {
                    // File nahi milti to URL ke bahawe try karo
                    player.setDataSource(path);
                }
            }
            
            player.prepareAsync();
            player.setOnPreparedListener(mp -> {
                mp.start();
                if (h.btnPlayPause != null) h.btnPlayPause.setImageResource(R.drawable.ic_pause);
            });
            player.setOnCompletionListener(mp -> {
                playingPos = -1;
                if (h.btnPlayPause != null) h.btnPlayPause.setImageResource(R.drawable.ic_play);
                try { mp.release(); } catch (Exception ignored) {}
                player = null;
            });
            player.setOnErrorListener((mp, what, extra) -> {
                android.util.Log.e("AudioPlay", "Error: " + what + " extra: " + extra + " path: " + path);
                playingPos = -1;
                if (h.btnPlayPause != null) h.btnPlayPause.setImageResource(R.drawable.ic_play);
                return true;
            });
        } catch (Exception e) {
            android.util.Log.e("AudioPlay", "playAudioFromPath error: " + e.getMessage() + " path: " + path);
            if (player != null) { try { player.release(); } catch (Exception ignored) {} player = null; }
        }
    }

    // ──────────────────────────────────────────────────────────────
    // Long-press bottom sheet actions
    // ──────────────────────────────────────────────────────────────
    private void showActionBottomSheet(Context ctx, Message m) {
        if (actionListener == null) return;
        String[] options = {"Reply", "Copy", "Star", "Forward", "Delete"};
        new android.app.AlertDialog.Builder(ctx)
            .setItems(options, (d, which) -> {
                switch (which) {
                    case 0: actionListener.onReply(m);   break;
                    case 1: actionListener.onCopy(m);    break;
                    case 2: actionListener.onStar(m);    break;
                    case 3: actionListener.onForward(m); break;
                    case 4: actionListener.onDelete(m);  break;
                }
            }).show();
    }

    @Override
    public void onViewRecycled(@NonNull VH holder) {
        super.onViewRecycled(holder);
        if (holder.ivImage != null) Glide.with(holder.ivImage).clear(holder.ivImage);
    }

    // ──────────────────────────────────────────────────────────────
    // ViewHolder — covers all view IDs used in both item layouts
    // ──────────────────────────────────────────────────────────────
    static class VH extends RecyclerView.ViewHolder {
        TextView     tvMessage, tvTime, tvSenderName, tvFileName;
        ImageView    ivImage;
        TextView     tvStatus;   // tv_status in both item layouts
        LinearLayout llAudio, llFile;
        ImageButton  btnPlayPause;   // ImageButton in both item layouts
        ImageView    btnDownload;

        VH(@NonNull View v) {
            super(v);
            tvMessage    = v.findViewById(R.id.tv_message);
            tvTime       = v.findViewById(R.id.tv_time);
            tvSenderName = v.findViewById(R.id.tv_sender_name);
            ivImage      = v.findViewById(R.id.iv_image);
            tvStatus     = v.findViewById(R.id.tv_status);
            llAudio      = v.findViewById(R.id.ll_audio);
            btnPlayPause = v.findViewById(R.id.btn_play_pause);
            llFile       = v.findViewById(R.id.ll_file);
            tvFileName   = v.findViewById(R.id.tv_file_name);
            btnDownload  = v.findViewById(R.id.btn_download);
        }
    }
}

File: CallX2/feature-chat/src/main/java/com/callx/app/adapters/MessagePagingAdapter.kt
package com.callx.app.adapters

import android.content.Context
import android.content.Intent
import android.media.MediaPlayer
import android.view.*
import android.widget.*
import androidx.paging.PagingDataAdapter
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.callx.app.chat.R
import com.callx.app.models.Message
import com.callx.app.utils.FileUtils
import com.callx.app.utils.MediaCache
import java.text.SimpleDateFormat
import java.util.*

/**
 * MessagePagingAdapter — Paging 3 PagingDataAdapter for chat messages (Kotlin).
 *
 * Supports sent/received layout types, text, image, audio, file, and video.
 */
class MessagePagingAdapter(
    private val currentUid: String,
    private val isGroup: Boolean
) : PagingDataAdapter<Message, MessagePagingAdapter.VH>(DIFF) {

    interface ActionListener {
        fun onReply(m: Message)
        fun onDelete(m: Message)
        fun onEdit(m: Message)
        fun onReact(m: Message, emoji: String)
        fun onForward(m: Message)
        fun onStar(m: Message)
        fun onPin(m: Message)
        fun onCopy(m: Message)
        fun onInfo(m: Message)
    }

    companion object {
        private const val TYPE_SENT        = 1
        private const val TYPE_RECEIVED    = 2
        private const val TYPE_STATUS_SEEN = 3
        private const val TYPE_REEL_SEEN   = 4

        private val DIFF = object : DiffUtil.ItemCallback<Message>() {
            override fun areItemsTheSame(a: Message, b: Message) =
                a.messageId != null && a.messageId == b.messageId
            override fun areContentsTheSame(a: Message, b: Message) =
                a.messageId == b.messageId && a.text == b.text &&
                a.type == b.type && a.status == b.status &&
                a.timestamp == b.timestamp && a.edited == b.edited && a.deleted == b.deleted
        }
    }

    private val timeFmt = SimpleDateFormat("hh:mm a", Locale.getDefault())
    var actionListener: ActionListener? = null
    private var player: MediaPlayer? = null
    private var playingPos = -1

    override fun getItemViewType(position: Int): Int {
        val m = getItem(position) ?: return TYPE_RECEIVED
        return when (m.type) {
            "status_seen" -> TYPE_STATUS_SEEN
            "reel_seen"   -> TYPE_REEL_SEEN
            else          -> if (m.senderId == currentUid) TYPE_SENT else TYPE_RECEIVED
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH {
        val layout = when (viewType) {
            TYPE_SENT        -> R.layout.item_message_sent
            TYPE_STATUS_SEEN -> R.layout.item_status_seen_bubble
            TYPE_REEL_SEEN   -> R.layout.item_reel_seen_bubble
            else             -> R.layout.item_message_received
        }
        val v = LayoutInflater.from(parent.context).inflate(layout, parent, false)
        return VH(v)
    }

    override fun onBindViewHolder(h: VH, position: Int) {
        val m = getItem(position) ?: return
        val sent = m.senderId == currentUid

        // Handle deleted messages
        if (m.deleted == true) {
            h.tvMessage?.text = "🚫 This message was deleted"
            h.tvMessage?.alpha = 0.6f
            bindFooter(h, m, sent)
            return
        }

        when (m.type) {
            "text", null -> bindText(h, m, sent)
            "image"      -> bindImage(h, m)
            "video"      -> bindVideo(h, m)
            "audio"      -> bindAudio(h, m, position)
            "file"       -> bindFile(h, m)
            else         -> bindText(h, m, sent)
        }

        // Sender name (group received messages)
        if (isGroup && !sent) {
            h.tvSenderName?.apply {
                text = m.senderName ?: ""
                visibility = if (!m.senderName.isNullOrEmpty()) View.VISIBLE else View.GONE
            }
        }

        // Reply preview
        if (!m.replyToId.isNullOrEmpty()) {
            h.llReplyPreview?.visibility = View.VISIBLE
            h.tvReplySender?.text = m.replyToSenderName ?: ""
            h.tvReplyText?.text = m.replyToText ?: "[message]"
        } else {
            h.llReplyPreview?.visibility = View.GONE
        }

        // Edited label
        h.tvEdited?.visibility = if (m.edited == true) View.VISIBLE else View.GONE

        // Reactions
        val reactStr = m.reactions?.values?.groupingBy { it }?.eachCount()
            ?.entries?.joinToString(" ") { "${it.key} ${it.value}" } ?: ""
        if (reactStr.isNotEmpty()) {
            h.llReactions?.visibility = View.VISIBLE
            h.tvReactions?.text = reactStr
        } else {
            h.llReactions?.visibility = View.GONE
        }

        bindFooter(h, m, sent)
        setupLongPress(h, m, sent, h.itemView.context)
    }

    private fun bindText(h: VH, m: Message, sent: Boolean) {
        h.tvMessage?.apply { text = m.text ?: ""; visibility = View.VISIBLE }
        h.ivImage?.visibility = View.GONE
        h.flVideo?.visibility = View.GONE
        h.llAudio?.visibility = View.GONE
        h.llFile?.visibility  = View.GONE
    }

    private fun bindImage(h: VH, m: Message) {
        h.tvMessage?.visibility = View.GONE
        h.ivImage?.apply {
            visibility = View.VISIBLE
            Glide.with(context).load(m.mediaUrl).placeholder(R.drawable.ic_gallery).into(this)
            setOnClickListener { openMedia(context, m.mediaUrl, "image") }
        }
    }

    private fun bindVideo(h: VH, m: Message) {
        h.tvMessage?.visibility = View.GONE
        h.flVideo?.visibility = View.VISIBLE
        h.ivVideoThumb?.let { iv ->
            val thumb = m.thumbnailUrl ?: m.mediaUrl
            Glide.with(iv.context).load(thumb).placeholder(R.drawable.ic_video).into(iv)
            h.flVideo?.setOnClickListener { openMedia(iv.context, m.mediaUrl, "video") }
        }
        if (m.duration != null) {
            h.tvVideoDuration?.apply { text = formatDuration(m.duration!!); visibility = View.VISIBLE }
        }
    }

    private fun bindAudio(h: VH, m: Message, pos: Int) {
        h.tvMessage?.visibility = View.GONE
        h.llAudio?.visibility = View.VISIBLE
        h.tvAudioDur?.text = if (m.duration != null) formatDuration(m.duration!!) else ""
        h.btnPlayAudio?.setOnClickListener { togglePlay(h, m, pos) }
    }

    private fun bindFile(h: VH, m: Message) {
        h.tvMessage?.visibility = View.GONE
        h.llFile?.visibility = View.VISIBLE
        h.tvFileName?.text = m.fileName ?: "File"
        h.tvFileMeta?.text = if (m.fileSize != null) "${m.fileSize!! / 1024} KB" else ""
        h.ivDownload?.setOnClickListener {
            m.mediaUrl?.let { url ->
                val intent = Intent(Intent.ACTION_VIEW, android.net.Uri.parse(url))
                h.itemView.context.startActivity(intent)
            }
        }
    }

    private fun bindFooter(h: VH, m: Message, sent: Boolean) {
        h.tvTime?.text = if (m.timestamp != null) timeFmt.format(Date(m.timestamp!!)) else ""
        h.tvStarredIcon?.visibility = if (m.starred == true) View.VISIBLE else View.GONE

        if (h.tvStatus != null) {
            if (sent) {
                h.tvStatus.visibility = View.VISIBLE
                when (m.status ?: "sent") {
                    "read"      -> { h.tvStatus.text = "✓✓ "; h.tvStatus.setTextColor(0xFFFF00FF.toInt()) }
                    "delivered" -> { h.tvStatus.text = "✓✓"; h.tvStatus.setTextColor(0xFFFF00FF.toInt()) }
                    else        -> { h.tvStatus.text = "✓"; h.tvStatus.setTextColor(0xCCFFFFFF.toInt()) }
                }
            } else {
                h.tvStatus.visibility = View.GONE
            }
        }
    }

    private fun setupLongPress(h: VH, m: Message, sent: Boolean, ctx: Context) {
        h.itemView.setOnLongClickListener {
            showActionSheet(ctx, m, sent)
            true
        }
    }

    private fun showActionSheet(ctx: Context, m: Message, sent: Boolean) {
        val sheet = com.google.android.material.bottomsheet.BottomSheetDialog(ctx)
        val sv = LayoutInflater.from(ctx).inflate(R.layout.bottom_sheet_message_actions, null)

        val emojis  = listOf("❤️", "👍", "😂", "😮", "😢", "🙏")
        val emojiIds = listOf(R.id.emoji_heart, R.id.emoji_thumb, R.id.emoji_laugh, R.id.emoji_wow, R.id.emoji_sad, R.id.emoji_pray)
        emojis.zip(emojiIds).forEach { (emoji, id) ->
            sv.findViewById<TextView>(id)?.setOnClickListener { sheet.dismiss(); actionListener?.onReact(m, emoji) }
        }

        sv.findViewById<View>(R.id.action_reply).setOnClickListener  { sheet.dismiss(); actionListener?.onReply(m) }
        sv.findViewById<View>(R.id.action_star).setOnClickListener   { sheet.dismiss(); actionListener?.onStar(m) }
        sv.findViewById<View>(R.id.action_pin).setOnClickListener    { sheet.dismiss(); actionListener?.onPin(m) }
        sv.findViewById<View>(R.id.action_forward).setOnClickListener { sheet.dismiss(); actionListener?.onForward(m) }

        if (sent && m.type == "text" && m.deleted != true) {
            sv.findViewById<View>(R.id.action_edit)?.let { it.visibility = View.VISIBLE; it.setOnClickListener { sheet.dismiss(); actionListener?.onEdit(m) } }
        }
        if (m.type == "text" && !m.text.isNullOrEmpty() && m.deleted != true) {
            sv.findViewById<View>(R.id.action_copy)?.let { it.visibility = View.VISIBLE; it.setOnClickListener { sheet.dismiss(); actionListener?.onCopy(m) } }
        }
        if (sent) {
            sv.findViewById<View>(R.id.action_info)?.let { it.visibility = View.VISIBLE; it.setOnClickListener { sheet.dismiss(); actionListener?.onInfo(m) } }
        }
        if (m.deleted != true) {
            sv.findViewById<View>(R.id.action_delete)?.setOnClickListener { sheet.dismiss(); actionListener?.onDelete(m) }
        }

        sheet.setContentView(sv); sheet.show()
    }

    private fun openMedia(ctx: Context, url: String?, type: String) {
        if (url.isNullOrEmpty()) return
        ctx.startActivity(Intent().apply {
            setClassName(ctx.packageName, "com.callx.app.activities.MediaViewerActivity")
            putExtra("url", url); putExtra("type", type)
        })
    }

    private fun formatDuration(ms: Long): String {
        val totalSec = ms / 1000
        return "%d:%02d".format(totalSec / 60, totalSec % 60)
    }

    private fun togglePlay(h: VH, m: Message, pos: Int) {
        if (m.mediaUrl == null) return
        if (player != null && playingPos == pos && player!!.isPlaying) {
            player!!.pause()
            h.btnPlayAudio?.setImageResource(R.drawable.ic_play)
            return
        }
        player?.let { try { it.release() } catch (_: Exception) {} }
        player = null; playingPos = pos
        h.btnPlayAudio?.setImageResource(R.drawable.ic_pause)

        val cached = MediaCache.getCached(h.itemView.context, m.mediaUrl!!)
        if (cached != null) playFromFile(h, cached.absolutePath, pos)
        else MediaCache.get(h.itemView.context, m.mediaUrl!!, object : MediaCache.Callback {
            override fun onReady(file: java.io.File) { playFromFile(h, file.absolutePath, pos) }
            override fun onError(reason: String)     { playFromFile(h, m.mediaUrl!!, pos) }
        })
    }

    private fun playFromFile(h: VH, path: String, pos: Int) {
        try {
            player = MediaPlayer()
            if (!path.startsWith("http")) {
                val f = java.io.File(path)
                if (f.exists()) java.io.FileInputStream(f).use { player!!.setDataSource(it.fd) }
                else player!!.setDataSource(path)
            } else {
                player!!.setDataSource(path)
            }
            player!!.prepareAsync()
            player!!.setOnPreparedListener { it.start() }
            player!!.setOnCompletionListener {
                h.btnPlayAudio?.setImageResource(R.drawable.ic_play)
                try { it.release() } catch (_: Exception) {}
                player = null; playingPos = -1
            }
        } catch (e: Exception) {
            player?.let { try { it.release() } catch (_: Exception) {} }
            player = null
        }
    }

    fun releasePlayer() {
        player?.let { try { it.release() } catch (_: Exception) {} }
        player = null; playingPos = -1
    }

    class VH(v: View) : RecyclerView.ViewHolder(v) {
        val tvMessage:     TextView?        = v.findViewById(R.id.tv_message)
        val tvTime:        TextView?        = v.findViewById(R.id.tv_time)
        val tvSenderName:  TextView?        = v.findViewById(R.id.tv_sender_name)
        val ivImage:       ImageView?       = v.findViewById(R.id.iv_image)
        val flVideo:       FrameLayout?     = v.findViewById(R.id.fl_video)
        val ivVideoThumb:  ImageView?       = v.findViewById(R.id.iv_video_thumb)
        val tvVideoDuration: TextView?      = v.findViewById(R.id.tv_duration)
        val llAudio:       LinearLayout?    = v.findViewById(R.id.ll_audio)
        val btnPlayAudio:  ImageButton?     = v.findViewById(R.id.btn_play_pause)
        val seekAudio:     SeekBar?         = v.findViewById(R.id.seek_audio)
        val tvAudioDur:    TextView?        = v.findViewById(R.id.tv_audio_dur)
        val llFile:        LinearLayout?    = v.findViewById(R.id.ll_file)
        val tvFileName:    TextView?        = v.findViewById(R.id.tv_file_name)
        val tvFileMeta:    TextView?        = v.findViewById(R.id.tv_file_meta)
        val ivDownload:    ImageButton?     = v.findViewById(R.id.btn_download)
        val tvStatus:      TextView?        = v.findViewById(R.id.tv_status)
        val llReplyPreview: LinearLayout?   = v.findViewById(R.id.ll_reply_preview)
        val tvReplySender:  TextView?       = v.findViewById(R.id.tv_reply_sender)
        val tvReplyText:    TextView?       = v.findViewById(R.id.tv_reply_text)
        val llReactions:   LinearLayout?    = v.findViewById(R.id.ll_reactions)
        val tvReactions:   TextView?        = v.findViewById(R.id.tv_reactions)
        val tvEdited:      TextView?        = v.findViewById(R.id.tv_edited)
        val tvPinnedLabel: TextView?        = v.findViewById(R.id.tv_pinned_label)
        val tvForwarded:   TextView?        = v.findViewById(R.id.tv_forwarded)
        val tvStarredIcon: TextView?        = v.findViewById(R.id.tv_starred_icon)
    }
}
